servers/republik/modules/crowdfundings/lib/scheduler/changeover.js

```diff
@@ -0,0 +1,146 @@
+const moment = require('moment')
+const Promise = require('bluebird')
+
+const {
+  resolveMemberships,
+  findDormantMemberships
+} = require('../CustomPackages')
+
+const createCache = require('../cache')
+
+const changeover = async (args, { pgdb, mail: { enforceSubscriptions } }) => {
+  const endDate = moment()
+
+  const users = await pgdb.public.query(`
+    SELECT "userId" AS id FROM (
+      -- Active memberships and their MAX(membershipPeriods.endDate)
+      SELECT m.*, MAX(mp."endDate") AS "endDate"
+      FROM memberships m
+      INNER JOIN "membershipPeriods" mp
+        ON m.id = mp."membershipId"
+      INNER JOIN "membershipTypes" mt
+        ON m."membershipTypeId" = mt.id
+        AND mt.name != 'MONTHLY_ABO' -- Exclude subscription via Stripe
+      WHERE
+        m.active = true
+      GROUP BY 1
+    ) AS users
+    WHERE
+      -- Potential ending memberships to change over to dormant membership
+      ( "endDate" < :endDate )
+    GROUP BY 1
+  `, { endDate })
+
+  debug({
+    now: endDate.toDate(),
+    users: users.length
+  })
+
+  const stats = { changeover: 0 }
+
+  await Promise.each(
+    users,
+    async user => {
+      // Find all memberships a user owns
+      let userMemberships = await pgdb.public.memberships.find(
+        { userId: user.id }
+      )
+
+      userMemberships = await resolveMemberships({
+        memberships: userMemberships,
+        pgdb
+      })
+
+      if (userMemberships.filter(m => m.active).length > 1) {
+        console.error(
+          'changeover failed: multiple active memberships found',
+          { user }
+        )
+        return
+      }
+
+      const activeMembership = userMemberships.find(m => m.active)
+      const dormantMemberships = findDormantMemberships({
+        memberships: userMemberships,
+        user: { id: user.id }
+      })
+
+      // End here if there is not dormant membership to be found
+      if (dormantMemberships.length < 1) {
+        return
+      }
+
+      // Elect a dormant membership to activate. Rule is to elect dorman
+      // membership with lowest sequenceNumber
+      const electedDormantMembership = dormantMemberships.reduce(
+        (acc, curr) =>
+          !acc || curr.sequenceNumber < acc.sequenceNumber
+            ? curr
+            : acc
+      )
+
+      const transaction = await pgdb.transactionBegin()
+      debug({
+        activeMembership: activeMembership.id,
+        electedDormantMembership: electedDormantMembership.id
+      })
+
+      try {
+        const now = moment()
+
+        await transaction.public.memberships.update(
+          { id: activeMembership.id },
+          { active: false, renew: false, updatedAt: now }
+        )
+
+        await transaction.public.memberships.updateOne(
+          { id: electedDormantMembership.id },
+          {
+            active: true,
+            renew: true,
+            sequenceNumber: activeMembership.sequenceNumber,
```

Comment:
I think it's a bad idea to copy the sequenceNumber over. First it's a unique field. Then It's used as a reference for humans all over the place. If you want to know what membership this new one succeeds, then add a `succeedsMembershipId` field and make the life of everybody who has to join this easier. Because we all know, if there's a reference we are going to use it. And we already have the code in place that always picks the lowest sequenceNumber for a given user.
If we want to free up membership sequenceNumbers in the future we should dedicate some thinking about all the sideeffects into it.